[Data Grid] Avoid `<GridRoot />` double-render pass on mount in SPA mode

lauri865 opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

<!-- You can use `## Changelog` to create a description for this change in the next release. -->

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/mui-x/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

Currently, GridRoot double-renders in all cases on mount to prevent SSR. In SPA-mode this is irrelevant, and can be avoided with the help of `useSyncExternalStore`, since there's no server snapshot in SPA-mode, the gird will be mounted directly. As a result, the rootRef becomes immediately available on first mount in SPAs, as one would expect.

Adds a new dependency `use-sync-external-store` to make it backwards compatible with React 17. Charts and treeview have the same dependency.

### Changelog

## Breaking changes

- The `filteredRowsLookup` object of the filter state does not contain `true` values anymore. If the row is filtered out, the value is `false`. Otherwise, the row id is not present in the object.
  This change only impacts you if you relied on `filteredRowsLookup` to get ids of filtered rows. In this case,use `gridDataRowIdsSelector` selector to get row ids and check `filteredRowsLookup` for `false` values:

  ```diff
   const filteredRowsLookup = gridFilteredRowsLookupSelector(apiRef);
  -const filteredRowIds = Object.keys(filteredRowsLookup).filter((rowId) => filteredRowsLookup[rowId] === true);
  +const rowIds = gridDataRowIdsSelector(apiRef);
  +const filteredRowIds = rowIds.filter((rowId) => filteredRowsLookup[rowId] !== false);
  ```

- The `visibleRowsLookup` state does not contain `true` values anymore. If the row is not visible, the value is `false`. Otherwise, the row id is not present in the object:

  ```diff
   const visibleRowsLookup = gridVisibleRowsLookupSelector(apiRef);
  -const isRowVisible = visibleRowsLookup[rowId] === true;
  +const isRowVisible = visibleRowsLookup[rowId] !== false;
  ```

lauri865 commented:
Side-benefit is also probably more robust tests if the Datagrid is mounted on the first pass.

romgrk commented on packages/x-data-grid/src/hooks/features/pagination/useGridPagination.ts:
```diff
@@ -29,7 +29,10 @@ export const paginationStateInitializer: GridStateInitializer<
   throwIfPageSizeExceedsTheLimit(paginationModel.pageSize, props.signature);
 
-  const rowCount = props.rowCount ?? props.initialState?.pagination?.rowCount;
+  const rowCount =
+    props.rowCount ??
+    props.initialState?.pagination?.rowCount ??
+    (props.paginationMode === 'client' ? state.rows?.totalRowCount : undefined);
```
Why was this change necessary?

lauri865 replied:
Because otherwise the pagination test breaks. Pagination is dependent on rowCount event for client-side pagination, however it is never explicitly initialized.

The problem doesn't surface with the current SSR blocking, because the state is implicitly initiated in a useEffect somewhere, before the grid can mount. 

Or that's how I understood it anyways.

romgrk replied:
Does that mean there is an occasion to remove a now redundant `useEffect`?

lauri865 replied:
Unfortunately not as a result of this change, this half could be changed if it could be tied to row updates for client-side pagination, but not sure if it's worth the effort. Since for server-side pagination, rowCount is not a derived state, but a standalone value that can be provided asynchronously, I'm not sure there's much to do about the useEffect:
https://github.com/mui/mui-x/blob/659dc5f42d3ba485c6dfdb7ae52f5e899bc66417/packages/x-data-grid/src/hooks/features/pagination/useGridRowCount.ts#L145-L147

Fwiw, this was the test that was breaking, was actually a cellSelection test, but pagination related still:
<img width="1488" alt="Screenshot 2024-12-02 at 20 33 32" src="https://github.com/user-attachments/assets/4c99de18-1fa9-4cd5-b124-1bff04b58e74">
<img width="1050" alt="Screenshot 2024-12-02 at 20 41 07" src="https://github.com/user-attachments/assets/64096103-75b3-4bba-b424-e6df7b5debf3">
<img width="1109" alt="Screenshot 2024-12-02 at 20 41 26" src="https://github.com/user-attachments/assets/09bf14a3-140d-4af8-9de6-5c9c6a0ebc71">

romgrk reviewed (approved):
Code LGTM, we can merge once the tests are passing.

lauri865 commented:
I have no idea how to fix the last failing test. Either it's a symptom of column headers updated more times now or just a bad test. But I lack understanding why it was previously expected that the warning message will be output exactly twice.

lauri865 commented:
Also fixes an annoying flicker coming out of loading state. It's probably the root cause of a layout shift I once reported with `autoHeight` grid coming out of loading state.

Before:
https://github.com/user-attachments/assets/1a68b2de-f2fd-4ef4-a93d-215d044e35b7

After:
https://github.com/user-attachments/assets/cfe55012-092a-429d-a79d-75e243124d10

lauri865 commented:
Tests should be passing now. 4 was the right amount of warnings for this approach, 1 extra pass (+1 warning for header/row each = 2+2 = 4) that doesn't even flush to the dom, so I think we're good there.

Curiously, after this optimisation:
https://github.com/mui/mui-x/blob/5ca918f1e949dac644168e7392706b933adc95bc/packages/x-data-grid/src/hooks/features/columns/useGridColumns.tsx#L400-L412

This test started failing:
https://github.com/mui/mui-x/blob/1a079e18fad1ddc2bf4a49f476bb5c68b47de1a2/packages/x-data-grid-premium/src/tests/aggregation.DataGridPremium.test.tsx#L145-L150

The only reason I can think of is that aggregations didn't explicitly trigger column updates, but somehow were piggybacking on `viewportInnerSizeChange` event for taking care of it. Adding an explicit `updateColumns` call fixed it. Is there any other explanation to it?

romgrk commented on packages/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx:
```diff
@@ -1104,3 +1104,7 @@ function bufferForDirection(
       throw new Error('unreachable');
   }
 }
+
+function roundToSubPixel(value: number) {
+  return Math.round(value * 10) / 10;
+}
```
Why is 1 decimal place a better value?

lauri865 replied:
Some browsers support subpixel positioning, but I saw some instances in the docs where width was 242.32px, which triggers unnecessary updates. Happy to change it to full pixels if that makes more sense.

romgrk commented on packages/x-data-grid-pro/src/tests/events.DataGridPro.test.tsx:
```diff
@@ -344,7 +344,14 @@ describe('<DataGridPro /> - Events params', () => {
       />,
     );
-    act(() => apiRef.current.publishEvent('scrollPositionChange', { left: 0, top: 3 * 52 }));
+    act(() => {
+      apiRef.current.publishEvent('renderedRowsIntervalChange', {
+        firstRowIndex: 3,
+        lastRowIndex: 6,
+        firstColumnIndex: 0,
+        lastColumnIndex: 0,
+      });
+    });
```
Why is this test modified?

lauri865 replied:
Seemed like a faulty premise for the test, but honestly, but in hindsight, I'm not sure why it previously worked to begin with.

Lazy loader only subscribes to `renderedRowsIntervalChange`:
https://github.com/mui/mui-x/blob/1a079e18fad1ddc2bf4a49f476bb5c68b47de1a2/packages/x-data-grid-pro/src/hooks/features/lazyLoader/useGridLazyLoader.ts#L134-L138

`scrollPositionChange` shouldn't change actual scroll position nor trigger `renderedRowsIntervalChange`, so how can it even trigger a new slice to be loaded?

Maybe you understand it better, but I tried to test it in the browser and everything worked fine with actual scrolling.

lauri865 replied:
This test gave me so much grief trying to trace in the code base how `scrollPositionChange` can possibly trigger lazyLoader. And guess what? It doesn't. 

Ran it on the mater branch and removing the `publishEvent` line will still make the pass. But for whatever reason, it will always try to load all the rows minus the first 3 that it has – tha args to onFetchRows are:
```
Object{firstRowToRender: 3, lastRowToRender: 500, sortModel: [], filterModel: Object{items: [], logicOperator: 'and', quickFilterValues: [], quickFilterLogicOperator: 'and'}}
```

Again, this is on the master branch, for the lack of any confusion.

lauri865 replied:
The test is still testing something useful that broke, just not what it says on the label currently. I reduced the amount `updateDimensions` calls, which called `updateRenderContext` ahead of dimensions state being updated, which prevented `lazyLoader` from working directly on mount (scroll still worked).

Fixed it now, and improved the tests – split it into two and improved the checks to more than just checking if fetchRows has been called.

romgrk commented on packages/x-data-grid-premium/src/hooks/features/aggregation/useGridAggregation.ts:
```diff
@@ -63,6 +63,7 @@ export const useGridAggregation = (
       const currentModel = gridAggregationModelSelector(apiRef);
       if (currentModel !== model) {
         apiRef.current.setState(mergeStateWithAggregationModel(model));
+        apiRef.current.updateColumns([]);
         apiRef.current.forceUpdate();
```
Why is this required?

While you're touching this, you could delete `.forceUpdate()`, it's a legacy function that should be removed.

lauri865 replied:
After some further debugging, it's not required as such, but there's a bug somehwere:
* Removing triggers `checkAggregationRulesDiff` -> triggers `hydrateColumns` -> updates columns (but aggregation is not cleared for whatever reason) -> triggers `viewportInnerSizeChange` -> triggers `handleGridSizeChange` (which often is unnecessary, unless there are flex columns) where aggregation is actually cleared.

So, it seems that `hydrateColumns` is probably ran before state has resolved.

The right fix then is to replace:
```typescript
useGridApiOptionHandler(apiRef, 'aggregationModelChange', checkAggregationRulesDiff);
```

With something along the lines of (unless there's a better internal way to chain this):
```
const aggregationModel = gridAggregationModelSelector(apiRef);
useEnhancedEffect(checkAggregationRulesDiff, [checkAggregationRulesDiff, aggregationModel]);
```

lauri865 replied:
Root cause is that these two get triggered out of order:
```typescript
useGridApiOptionHandler(apiRef, 'aggregationModelChange', checkAggregationRulesDiff);

React.useEffect(() => {
    if (props.aggregationModel !== undefined) {
      apiRef.current.setAggregationModel(props.aggregationModel);
    }
  }, [apiRef, props.aggregationModel]);
```

As soon as the prop changes, `checkAggregationRulesDiff` is triggered, ahead of `setAggregationModel`. And it currently only works, because column update always triggers two state updates to columns (which in most cases is not needed), which triggers a second hydration from aggregation, only at which point the aggregation state has been updated.

lauri865 replied:
cc @flaviendelangle, any thoughts? 🙏 Seems like you worked on this feature.

flaviendelangle replied:
I built the feature 2.5 years ago so my understanding of it might be rusted :laughing: 

> As soon as the prop changes, checkAggregationRulesDiff is triggered

That surprises me.
`checkAggregationRulesDiff` is triggered by the `aggregationModelChange` event, which is fired in the `setState` method, which is called by `setAggregationModel`

So basically:

`props.aggregationModel` changes => `setState` is called => ` state.aggregation.model` is updated and `aggregationModelChange` is fired => `checkAggregationRulesDiff` is called

Maybe we have a problem during the initialization though.

The problem with firing updated like `checkAggregationRulesDiff` in effects is that you might end up with inconsistent states and flickering, but I don't know if that's the case for this specific feature.

lauri865 replied:
There problem seems to be on initialization indeed. Not sure if it has any real-world bearing, but on first prop change, the values are as such:
```
aggregationRules {}
rulesOnLastColumnHydration {}
aggregationRules {}
areAggregationRulesEqual(rulesOnLastColumnHydration, aggregationRules)
```

`rulesOnLastColumnHydration` is empty, even though columns have rules applied as can be seen from the test result:
```
1) <DataGridPremium /> - Aggregation
       Setting aggregation model
         prop: aggregationModel
           should correctly restore the column when changing from aggregated to non-aggregated:

      AssertionError: expected 'idmax' to equal 'id'
      + expected - actual

      -idmax
      +id
```

lauri865 replied:
This hack also helps the test pass, but not sure what the actual fix here is:
```typescript
export const aggregationStateInitializer: GridStateInitializer<
  Pick<DataGridPremiumProcessedProps, 'aggregationModel' | 'initialState'>,
  GridPrivateApiPremium
> = (state, props, apiRef) => {
  apiRef.current.caches.aggregation = {
    rulesOnLastColumnHydration: props.aggregationModel !== undefined ? ({ _init: {} } as any) : {},
    rulesOnLastRowHydration: {},
  };

  return {
    ...state,
    aggregation: {
      model: props.aggregationModel ?? props.initialState?.aggregation?.model ?? {},
    },
  };
};
```

lauri865 replied:
aha, I think I see the actual fix now

lauri865 replied:
I think the right place to update `apiRef.current.caches.aggregation.rulesOnLastColumnHydration` is in the preprocessor after `hydrateColumns` and not in `checkAggregationRulesDiff`

romgrk commented on packages/x-data-grid/src/tests/layout.DataGrid.test.tsx:
The argos CI test has a few failing cases, some of which would need to be fixed (at least the "no rows" one).

lauri865 replied:
"no rows" one should be fixed now. Created a selector for pinnedRow heights, which improves `updateDimensions` flow of updates (imho).

lauri865 commented:
I'm starting to regret opening this can of worms 🙈

@romgrk, I'm hitting a very strange bug in regression tests – not sure if it's related to only that or overall, but this selector returns stale state, but only for the horizontal scrollbar:
https://github.com/mui/mui-x/blob/1a079e18fad1ddc2bf4a49f476bb5c68b47de1a2/packages/x-data-grid/src/components/virtualization/GridVirtualScrollbar.tsx#L78C4-L78C72

Things that seem to fix it:
```typescript
const dimensions = gridDimensionsSelector(apiRef.current.state);
```

Removing the check for `dimensions.hasScrollX`:
```
{dimensions.hasScrollX && !rootProps.unstable_listView && (
  <Scrollbar position="horizontal" {...getScrollbarHorizontalProps()} />
)}
```

Both vertical scrollbar as well as `GridVirtualScroller` have the correct state from the selector. I'm at my wits` end here.

lauri865 commented:
Ok, so it seems like `useGridSelector` can under certain circumstances return the stale state:
* Component mounts
* Selector sets initialState
* Something triggers a state change before Component can mount
* Store subscription is added too late, we don't catch the state update, and internal state of the selector remains stale.

I will change `useOnMount` to `useEnhancedEffect` with an empty dependency array, unless there are any objections @romgrk? `Store.subscribe` is a very simple operation, so shouldn't make any difference performance-wise.

lauri865 commented:
Tests passing 🎉
(this few line change turned out to be quite a can of worms, but all for the better I hope 🙈)

lauri865 commented:
@romgrk, made some changes and merged recent updates into this – ready to review and merge.

Managed to split up `useGridDimensions` and `useGridVirtualScroller` based on responsibility. Currently, dimensions hook triggers `renderContext` and always updates the `dimensions` object as well, even if it doesn't need any changing. This was difficult to read and trace. Now, `useGridDimensions` is only responsible for dimension related changes, and `useGridVirtualContext` handles its own events that require re-evaluating the `renderContext`. Much better imho, and less updates.

A couple of side-notes:
* As a result of this PR and some other recent changes, I believe throttling resize events can be disabled by default, as it doesn't seem to bring much performance-wise anymore. I suggest disabling it by default, but leaving in the prop if anyone needs it still. But for cases that actually need it, I think `debouce` might be better than `throtttle` (e.g. for smooth animations in viewport/container).
* `apiRef.current.forceUpdate` can be safely removed everywhere based on my testing. As a result of memoizing `GridRoot` and moving child components within that in a previous PR, the likelihood that it actually triggers any side-effects became much smaller as well, so less things to test / miss. But this can be moved to a different PR of course, or at least I'd wait until the other stuff is reviewed, to avoid spamming new changes.

romgrk commented:
> Edit: also deprecated forceUpdate as it serves no function – removed its functionality, happy to remove the invocations as well, but thought it would make the review very noisy.

All good, that function is a leftover from the previous reactivity model, we've been slowly removing it.

romgrk commented on packages/x-data-grid-premium/src/hooks/features/aggregation/useGridAggregationPreProcessors.tsx:
```diff
@@ -73,6 +73,7 @@ export const useGridAggregationPreProcessors = (
       });
 
       rulesOnLastColumnHydration.current = aggregationRules;
+      apiRef.current.caches.aggregation.rulesOnLastColumnHydration = aggregationRules;
 
       apiRef.current.caches.aggregation.rulesOnLastColumnHydration = aggregationRules;
```
Why is it repeated twice?

lauri865 replied:
Must be merge error / mistake. Fixed.

romgrk commented on packages/x-data-grid/package.json:
```diff
@@ -52,7 +52,8 @@
     "prop-types": "^15.8.1",
-    "reselect": "^5.1.1"
+    "reselect": "^5.1.1",
+    "use-sync-external-store": "^1.0.0"
```
Sidenote: if we're adding this dependency now, we could replace our `Store`/`useGridSelector` implementation with it. I don't remember why I didn't use it back then.

romgrk commented on packages/x-data-grid/src/components/GridRow.tsx:
```diff
@@ -56,7 +56,6 @@ export interface GridRowProps extends React.HTMLAttributes<HTMLDivElement> {
    */
   index: number;
   rowHeight: number | 'auto';
-  offsetTop: number | undefined;
```
I think I left this prop there to be able to play with the CSS layout method (e.g. switch to `transform: translate()` on each row) without introducing a breaking change (this slot is exposed publicly), but I guess it's fine to remove it as we don't plan to do any change.

romgrk commented on packages/x-data-grid/src/hooks/features/dimensions/useGridDimensions.ts:
```diff
@@ -79,12 +76,54 @@ const EMPTY_DIMENSIONS: GridDimensions = {
 };
 
-export const dimensionsStateInitializer: GridStateInitializer<RootProps> = (state) => {
+const getStaticDimensions = (
```
lint: prefer `function`, and keep `*/hooks/features/*` files organized as such:

```
// imports & types

const xxxStateInitalizer = () => { ... }

function useXxx() { ... }

// helper functions
```

lauri865 replied:
Good point, fixed.

lauri865 commented:
@romgrk, can you please review this last change as well please? Small performance improvement, but with large data sets can shave a few ms off – autogenerated row cannot exist with the same `id` has a regular row, right?
https://github.com/mui/mui-x/pull/15648/commits/a66231d6d0088abd03e7082843fa6c4534769df2

romgrk commented on packages/x-data-grid/src/hooks/features/filter/useGridFilter.tsx:
```diff
@@ -456,7 +454,9 @@ export const useGridFilter = (
-        filteredRowsLookup[id] = isRowPassing;
+        if (!isRowPassing) {
+          filteredRowsLookup[id] = isRowPassing;
+        }
```
I seem to remember that lookup object had weird semantics, `true`, `false` and `undefined` each had a meaning, is that still the case?

This is also another use-case where we should really switch to `Set` objects.

lauri865 replied:
I only saw on implementation that checked for a `true` value, which I changed. `undefined` should not have a meaning from what I can gather. If we can lose `true` value (which seems possible based on my testing manual and unit), then skipping unnecessary loops when filters are not present becomes much easier.

`Map` could make sense, but after this change the impact is gonna be quite a bit smaller, but still sensible to do it later on. But I'd rather skip it within this PR for now.

romgrk commented on packages/x-data-grid/src/hooks/utils/useGridSelector.ts:
```diff
@@ -1,11 +1,11 @@
 import * as React from 'react';
+import { unstable_useEnhancedEffect as useEnhancedEffect } from '@mui/utils';
```
Import directly from `@mui/utils/useEnhancedEffect`.

lauri865 replied:
I did initially, but for some reason every other import in the datagrid package has this one instead, any idea why?

romgrk replied:
Multiple generations of maintainers, but direct import is better for devtime HMR and has less chance of breaking brittle tree-shaking implementations, so new code should use that pattern.

lauri865 replied:
It shouldn't make a difference so long as there's at least one indirect import. Depending on tooling, but import paths get cached.

Default imports are yuck though, almost made a typo there already 🙈:
`import useEnhancedEffect from '@mui/utils/useEnhancedEffect';`

romgrk commented on packages/x-data-grid/src/hooks/utils/useGridSelector.ts:
```diff
@@ -102,7 +102,7 @@ export const useGridSelector = <Api extends GridApiCommon, Args, T>(
     }
   }
 
-  useOnMount(() => {
+  useEnhancedEffect(() => {
```
The thing I like about `useOnMount` is that it uses a `const EMPTY = []` array for its dependencies, so it doesn't allocate anything when it runs. `useGridSelector` is used pervasively in components (and in rows/cells in particular), so during high-speed scrolling this new `[apiRef, refs]` array could easily generate hundreds or thousands of needless memory allocations, which I constantly try to get rid of due to perf issues.

I would repeat the `const EMPTY = []` here, and mute the eslint rule that's going to complain about `apiRef` and `refs`, those are stable ref objects.

lauri865 replied:
Makes sense. Done.

romgrk commented on packages/x-data-grid/src/hooks/features/sorting/gridSortingSelector.ts:
```diff
@@ -33,10 +33,11 @@ export const gridSortedRowEntriesSelector = createSelectorMemoized(
-      const rowNode = rowTree[id];
-      if (rowNode && isAutogeneratedRowNode(rowNode)) {
-        acc.push({ id, model: { [GRID_ID_AUTOGENERATED]: id } });
+      } else {
+        const rowNode = rowTree[id];
+        if (rowNode && isAutogeneratedRowNode(rowNode)) {
+          acc.push({ id, model: { [GRID_ID_AUTOGENERATED]: id } });
+        }
```
> @romgrk, can you please review this last change as well please? Small performance improvement, but with large data sets can shave a few ms off – autogenerated row cannot exist with the same id has a regular row, right?

There's a soft guarantee of that, autogenerated ids are in the format `auto-generated-row-${id}`, but I think we can accept that we reserve that pattern for us. Would have been nice to vendor-prefix it like we do for other identifiers though, or use symbols.

cherniavskii replied:
> I think we can accept that we reserve that pattern for us.

Agreed 👍🏻

> or use symbols.

Using `Symbol.for`?

```diff
- `auto-generated-row-${pathStr}`
+ Symbol.for(`auto-generated-row-${pathStr}`)
```

lauri865 replied:
If I remember correctly, `Symbol.for` persists in memory for the whole session, and never get garbage collected / can't be cleaned up manually.

romgrk replied:
Yes indeed. I think a reserved string pattern is enough. I would be happier if it was `mui:...${id}`, but I think the `auto-generated-row-...` pattern has been publicized in the docs so changing it could break someone's code.

lauri865 commented:
@romgrk, added one last change™, to aide with #16372 (was critical for us as well) – using granular selectors for dimensions (in most places).  We're getting buttery smooth animations without throttling the resize now in our app (`.

Not sure where to drop all these selectors though, if we're fine with keeping them there, just need to run `pnpm docs:api` and we're golden:
https://github.com/mui/mui-x/pull/15648/commits/c2f90cbb9ab3159e9ad2ba2ee9b448910be88b5d#diff-2463b4ef2c8e9ed6ae758ab8d811375bc6c7cb7a8a384203c6ae9481ed48a3ca

romgrk commented on packages/x-data-grid/src/hooks/features/dimensions/gridDimensionsSelectors.ts:
```diff
@@ -1,3 +1,33 @@
+ */
+export const gridRowHeightSelector = (state: GridStateCommunity) => state.dimensions.rowHeight;
+
+export const gridDimensionsColumnsTotalWidthSelector = (state: GridStateCommunity) =>
```
"Dimensions" could be removed, it's clear what it does without it.

lauri865 replied:
there is a conflicting selector though that calculates it based on state 🤔

romgrk replied:
Does it return the same value? Should it be eliminated in favor of this one? If not, this one could be just `columnsTotalWidthSelector` considering it's going to be internal only.

romgrk commented:
> added one last change™

Yes, it LGTM but please if you have more, separate PR, I'm always nervous merging large PRs :sweat_smile:

> Edit: it's been ages I used anything emotion/styled, but isn't re-creating this gridStyle object on every render problematic?

Using emotion at all is problematic. Re-creating the style object is bad for memory allocations but the grid rarely has more than 1-2 instances per page so that's not much of an issue. Preserving the style object doesn't help emotion though, because emotion re-serializes each style object on each render, no matter if it has already seen it or not. To prevent emotion from doing that, we need to pass the object through emotion's underlying serializer and feed that output into emotion, like I did in https://github.com/mui/material-ui/pull/43412 for the core. I haven't had time to port the optimizations I did in the core to the grid, but we could maybe re-use `memoTheme` from that PR.

lauri865 commented:
> Yes, it LGTM but please if you have more, separate PR, I'm always nervous merging large PRs 😅
I'm sorry 🙈

> Using emotion at all is problematic. Re-creating the style object is bad for memory allocations but the grid rarely has more than 1-2 instances per page so that's not much of an issue. Preserving the style object doesn't help emotion though, because emotion re-serializes each style object on each render, no matter if it has already seen it or not. To prevent emotion from doing that, we need to pass the object through emotion's underlying serializer and feed that output into emotion, like I did in [mui/material-ui#43412](https://github.com/mui/material-ui/pull/43412) for the core. I haven't had time to port the optimizations I did in the core to the grid, but we could maybe re-use `memoTheme` from that PR.

Why is it re-evaluating on each resize though for `DatagridPro`? 🫣 If I remove the `Watermark` component it seems to stop doing that.

lauri865 commented:
@romgrk, this seems pretty critical – as long as `GridRoot` has any children at all (memoized or not), it re-evaluates the styles on each render. If I remove the children, it stops doing that – `DatagridPro` and `DatagridPremium` seem to be affected. `GridRootStyles` can have any children, as long as it's not passed from above. I really don't follow this.

lauri865 commented:
The below seems to fix it, still no idea why it's happening though:
```typescript
const watermark = React.useMemo(
    () => <Watermark packageName="x-data-grid-pro" releaseInfo={releaseInfo} />,
    [],
  );

  return (
    <GridContextProvider privateApiRef={privateApiRef} configuration={configuration} props={props}>
      <GridRoot
        className={props.className}
        style={props.style}
        sx={props.sx}
        {...props.slotProps?.root}
        ref={ref}
      >
        {watermark}
      </GridRoot>
    </GridContextProvider>
  );
```

romgrk commented:
> this seems pretty critical – as long as GridRoot has any children at all (memoized or not), it re-evaluates the styles on each render

What do you mean by critical? I know that re-evaluating the styles is bad, but is it measurably bad?

> The below seems to fix it, still no idea why it's happening though

Seems clear to me that `GridRoot`'s [memo](https://github.com/mui/mui-x/blob/816ca21402b0ba45095a6baba5311295b4b5f945/packages/x-data-grid/src/components/containers/GridRoot.tsx#L102) can't work if there is an unstable prop, and re-rendering `GridRoot` means re-rendering its main child, `GridRootStyles`. I'm happy to [think of the children](https://attardi.org/why-we-memo-all-the-things/#will-someone-please-think-of-the-children), though any other unstable prop would also break the memo.

romgrk commented on packages/x-data-grid-premium/src/DataGridPremium/DataGridPremium.tsx:
```diff
@@ -47,6 +47,12 @@ const DataGridPremiumRaw = forwardRef(function DataGridPremium<R extends GridVal
     validateProps(props, dataGridPremiumPropValidators);
   }
+
+  const watermark = React.useMemo(
+    () => <Watermark packageName="x-data-grid-premium" releaseInfo={releaseInfo} />,
+    [],
+  );
```
I think the watermark can be a constant outside of the render function.

lauri865 replied:
Good. Done.

lauri865 commented:
> > this seems pretty critical – as long as GridRoot has any children at all (memoized or not), it re-evaluates the styles on each render
> 
> What do you mean by critical? I know that re-evaluating the styles is bad, but is it measurably bad?
> 
> > The below seems to fix it, still no idea why it's happening though
> 
> Seems clear to me that `GridRoot`'s [memo](https://github.com/mui/mui-x/blob/816ca21402b0ba45095a6baba5311295b4b5f945/packages/x-data-grid/src/components/containers/GridRoot.tsx#L102) can't work if there is an unstable prop, and re-rendering `GridRoot` means re-rendering its main child, `GridRootStyles`. I'm happy to [think of the children](https://attardi.org/why-we-memo-all-the-things/#will-someone-please-think-of-the-children), though any other unstable prop would also break the memo.

Fair. I'm fine if it re-evaluates the styles on `rootProp` changes, but if it re-evaluates on every frame when I resize the grid, that might be a bit more problematic. Anyways, fine to merge from my end, promise not to add anything else here.

romgrk commented on packages/x-data-grid/src/hooks/features/dimensions/useGridDimensions.ts:
```diff
@@ -397,6 +376,24 @@ export function useGridDimensions(apiRef: RefObject<GridPrivateApiCommunity>, pr
 
+function setCSSVariables(root: HTMLElement, dimensions: GridDimensions) {
+  const set = (k: string, v: string) => root.style.setProperty(k, v);
+  // @deprecated due to extensive dom updates
+  // set('--DataGrid-width', `${dimensions.viewportOuterSize.width}px`);
```
No need to deprecate, this isn't public, we can remove it directly.

lauri865 commented:
@cherniavskii, these two tests started failing after I merged in the latest `master` branch:
https://app.circleci.com/pipelines/github/mui/mui-x/79790/workflows/07d38763-f916-416e-b4d9-87f3d612b609/jobs/454429

If you have any hunch why, would appreciate if you could take a look 🙏 Otherwise, will take a look later.

lauri865 commented:
@cherniavskii, seems like the debounce added in #16101 made the test flaky. Fails intermittently locally as well, but more consistently in CI, but not always. Seen it elsewhere as well now, so not particular to this PR it seems. The fetching probably needs an `AbortSignal` as well, otherwise you can end up with race conditions, and results coming in after unmounts.

Two alternatives to debounce:
1. Simply limit one call per render cycle
2. Serialize and store last fetch params to avoid refetching when params didn't actually change (could happen when clearing filters twice in a row)

lauri865 commented:
@romgrk, ready to merge now that I fixed the flaky tests. Also added one last optimisation – `hydrateRowsMeta` was called too many times due to unnecessary selectors (main culprit in this case was pagination model selector). I guess it was necessary in the past when `useGridVisibleRows` was not really reactive, but now it just adds extra renders due to `pagination.rowCount` updating for example after rows have already been updated. Filters, sorting, pagination, etc. propagates directly through `useGridVisibleRows`, so no need for extra logic / state there.

https://github.com/mui/mui-x/pull/15648/commits/ffa1efcc96d8993c319d62094274c7aa16d785f2

cherniavskii commented on packages/x-data-grid/src/hooks/utils/useIsSSR.ts:
I noticed this weird behavior in some demos: https://deploy-preview-15648--material-ui-x.netlify.app/x/react-data-grid/#community-version


https://github.com/user-attachments/assets/cdc7ace7-cbc3-4423-9d5a-f74f193a83cc

Looks like it only happens when the initial render is on the server (Argos did not catch this issue since we only use client rendering for screenshots).
I'll look into it.

lauri865 replied:
Thanks! Seems to be a timing issue with `useGridSelector`. Dimensions get initialised in a ref callback, which runs before `useLayoutEffect` and after the body of the hook. Subscription is added after the state is set, which doesn't catch the update. I guess the root cause always existed, it's just that extraneous state updates masked it.

lauri865 replied:
It works with `useInsertionEffect`. But needs a fallback for React 17, as it only exists from React 18.

cherniavskii replied:
Hmm, shouldn't we use `useSyncExternalStore` for subscription here?

lauri865 replied:
We could fallback to `useLayoutEffect` in React 17, but also need to call `applySelector` once more before subscription:
https://www.npmjs.com/package/@emotion/use-insertion-effect-with-fallbacks

lauri865 replied:
> Hmm, shouldn't we use `useSyncExternalStore` for subscription here?

Good point. That's even better.

cherniavskii replied:
> It works with useInsertionEffect

Partially. Column headers are still off, hovering over rows does nothing, etc.

<img width="839" src="https://github.com/user-attachments/assets/358d6e1a-2cfd-4409-85d9-6732dcda746c" />

I tried `useSyncExternalStore` instead, but the behavior is the same as with `useInsertionEffect`.

lauri865 replied:
Strange, the headers work fine on my end. What browser?

lauri865 replied:
Tried Edge, Chrome, Safari, Firefox – all work on Mac at least.

lauri865 replied:
Ok, managed to reproduce it – had to switch to light mode. No idea what's causing it though. Dark mode works perfectly.

cherniavskii replied:
This is so weird 🤔

lauri865 replied:
Fixed :)

lauri865 commented:
> @romgrk, ready to merge now that I fixed the flaky tests. Also added one last optimisation – `hydrateRowsMeta` was called too many times due to unnecessary selectors (main culprit in this case was pagination model selector). I guess it was necessary in the past when `useGridVisibleRows` was not really reactive, but now it just adds extra renders due to `pagination.rowCount` updating for example after rows have already been updated. Filters, sorting, pagination, etc. propagates directly through `useGridVisibleRows`, so no need for extra logic / state there.
> 
> [ffa1efc](https://github.com/mui/mui-x/commit/ffa1efcc96d8993c319d62094274c7aa16d785f2)
> 
> Should further improve mounting performance, since it removes a: `useLayoueEffect (setState)` -> `useEffect(setState)` -> `useLayoutEffect(setState)` loop.

Reg `useRowsMeta`, there's a bit more space for optimisation (for reference for the future, not gonna do it here):
Rows are set + filters, sorting, etc. resolved -> `useGridSelector` to get the `useGridVisibleRows` (triggers new render) -> `useEnhancedEffect` to `hydrateRowsMeta` (triggers new render) -> `useGridSelector` in `useGridDimensions` to subscribe to `rowsMeta` -> `useEnhancedEffect` to `updateDimensions` -> finally can calculate `renderContext` (in the same pass as `updateDimensions`. While it doesn't end up painting the DOM in the process, there's still a lot of diffing and processing happening there.

This, what is essentially a state reducer, could all happen in one rendering phase.

`useGridSelectorEffect` could make it a bit easier to write in a decoupled way – essentially subscribing to the state store based on selectors, and manipulating state based on whether a slice of the global state that interests us changed. Alternatively, `hydrateRowsMeta` could emit `rowsMetaChanged`, that `useGridDimensions` can subscribe to, but that's only half the story. The first part is more tricky with events, as `visibleRows` doesn't strictly live in the state store, but an output of different parts of the state.

lauri865 commented:
@cherniavskii and @romgrk, can you please review my `useGridSelector` implementation with `useSyncExternalStore`. I'm fine also putting it in a separate PR, but this one depends on it.

The only issue, which is currently also present in the current implementation is that the below only works due to side-effects of other state changing. If you only change the selector, and nothing else changes, then the state will be stale. I'm not sure we want to even support such selectors (looks a bit iffy to me), but worth keeping in mind:

https://github.com/mui/mui-x/blob/0736d31b9dc3a7197996317f010b76e68f3d3285/packages/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx#L99-L103

lauri865 commented:
Bad news on the `useSyncExternalStore` front – it doesn't support state batching like `useState`, so it's probably a dead end: https://github.com/facebook/react/issues/25191#issuecomment-1237456448

lauri865 commented:
Committed an alternative approach – using `useSyncExternalStore` to subscribe / unsubscribe, but managing the state with `useState` still in order to benefit from batching. Clipboard tests still break for some reason.

cherniavskii commented on packages/x-data-grid/src/hooks/utils/useGridSelector.ts:
```diff
@@ -102,20 +106,43 @@ export const useGridSelector = <Api extends GridApiCommon, Args, T>(
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, EMPTY);
+
+  useSyncExternalStore(unsubscribe, subscribe, emptyGetSnapshot);
```
Isn't it overcomplicated?
I thought this would work:

```tsx
  const getSnapshot = () => {
    const newState = applySelector(
      apiRef,
      refs.current.selector,
      refs.current.args,
      apiRef.current.instanceId,
    ) as T;
    if (!refs.current.equals(refs.current.state, newState)) {
      refs.current.state = newState;
      setState(newState);
    }
  };

  useSyncExternalStore(apiRef.current.store.subscribe, getSnapshot, getSnapshot);
```

lauri865 replied:
The subscription function runs at the same time as `useLayoutEffect`, so it doesn't fix the timing issue. And we don't actually want to use the snapshot nor the actual `useSyncExternalStore` subscription callback, as it triggers a synchronous re-render whenever it changes. Plus snapshot functions need to be stable, otherwise they run on every render. State batching is much better

So, instead, we hook into getSnapshot on first invocation to subscribe synchronously, and use the subscription cleanup function to tear it down. It's a hack, but hey, at least it seems to work with StrictMode as well.

romgrk replied:
Considering those points, do you think using `useSyncExternalStore` is better than the previous implementation with just `useLayoutEffect`?

lauri865 replied:
It's not that I think it's a better solution, but the only one that a) solves the timing issue b) works with StrictMode c) offers out of the box backwards compatibility.

The only other options that solve the timing issue are:
1) `useFirstRender` – doesn't jive well with StrictMode
2) `useInsertionEffect` – doesn't support React 17
3) somehow manually delaying state updates that happen in ref callback phase (when elements mount in the DOM, before `useLayoutEffect`

lauri865 replied:
I'm very open to alternatives approaches that make sure we don't end up with stale state, definitely not in love with the `useSyncExternalStore` hack 🙈

cherniavskii replied:
It seems to work fine.
Instead of subscribing in the first callback, we only use it for cleanup.
Then in the initial `getSnapshot` call we subscribe to the store and all updates go via state updates.
`getSnapshot` always returns `null`, so `useSyncExternalStore` doesn't cause direct rerenders (only indirect via state update).
I'm OK with proceeding with this solution.

cherniavskii commented:
> Edit: accidentally added back this line – lesson to everyone, avoid frequently changing CSS variables, it cut our FPS in half in our animation stress test:
set('--DataGrid-width', ${dimensions.viewportOuterSize.width}px);

Yeah, I was surprised to see you added it back 😅

cherniavskii commented on docs/data/data-grid/filtering-recipes/FilteredRowCount.js:
I agree with the proposed change, but why doesn't the previous implementation work anymore?

lauri865 replied:
Due to this change:
https://github.com/mui/mui-x/blob/cddb1bf4e08fda0cccfa8efdb5a9c77a3cdaca43/packages/x-data-grid/src/hooks/features/filter/useGridFilter.tsx#L457-L459

If we can skip `filteredRowsLookup` assigning values for both `true` and `false`, we can easily skip many `O(n)` operations (e.g. building `filteredRowsLookup` to begin with, but there was more cascading ones) if there are no filters present, as well have a smaller resulting object. Then we can just check for `isObjectEmpty` and skip the ops. There was only once instance internally that checked for `true` value and everything else checked for `filteredRowsLookup[id] !== false`.

lauri865 replied:
E.g. editing cells in this demo should be a few frames faster now when you commit the changes (same for initial mount):
https://deploy-preview-15648--material-ui-x.netlify.app/x/react-data-grid/#pro-version

cherniavskii replied:
Makes sense 👍🏻
I'll mention this change in the changelog, and perhaps revert it in a v7 cherry-pick PR to avoid breaking things.

lauri865 commented:
A bit offtopic, as I don't plan to add it to this PR, but also devised a small optimisation for scroll performance now that I looked a bit more into selectors. Selector performance seems pretty optimised, and they work great overall. But there's one part of the global state that can update very rapidly – `renderContext`. And the number of store listeners grows quite a bit the bigger the viewport, adding for more processing of selectors that don't actually care about the `renderContext`. In reality, there's only 3 components that are reactive consumers of the `renderContext`, so I would suggest breaking it out of the global store, and using a simple `useState` instead in `useGridVirtualScroller`; we can still update the global store, just not trigger `store.update`. The other two places can subscribe to the changes with event handlers and create their own slice of the state, or we could add `useGridRenderContext` hook to subscribe to that.

Before:

https://github.com/user-attachments/assets/4f36b26f-d3b8-40b2-81d2-53783229e2d8

After:

https://github.com/user-attachments/assets/094bfad9-1d2d-47e7-9b06-f3e6a711e5e7

___

Before (larger viewport):

https://github.com/user-attachments/assets/77140265-4f61-4c13-b87c-c35154a5abf2

After (larger viewport, with `flushSync` removed):

https://github.com/user-attachments/assets/daf614e2-0f3a-44c8-b7c5-d61d82da4517

(it's not exactly representative of my experience as I'm on a 120hz screen, and the video is converted to 60fps as I had no other options in the recording software, so there's some skipped frames)

romgrk commented:
> Edit: accidentally added back this line – lesson to everyone, avoid frequently changing CSS variables, it cut our FPS in half in our animation stress test: `set('--DataGrid-width', `${dimensions.viewportOuterSize.width}px`);`

We had a [similar case during scrolling](https://github.com/mui/mui-x/pull/12019), but I didn't think to apply the same change to resizing.

romgrk commented:
> A bit offtopic, as I don't plan to add it to this PR, but also devised a small optimisation for scroll performance now that I looked a bit more into selectors. Selector performance seems pretty optimised, and they work great overall. But there's one part of the global state that can update very rapidly – `renderContext`. And the number of store listeners grows quite a bit the bigger the viewport, adding for more processing of selectors that don't actually care about the `renderContext`. In reality, there's only 3 components that are reactive consumers of the `renderContext`, so I would suggest breaking it out of the global store, and using a simple `useState` instead in `useGridVirtualScroller`; we can still update the global store, just not trigger `store.update`. The other two places can subscribe to the changes with event handlers and create their own slice of the state, or we could add `useGridRenderContext` hook to subscribe to that.

I've also observed that issue, and I've mentionned that we could get rid of the perf cost of running all the selectors during scrolling by switching from selector-based reactivity to event-based reactivity ([conceptual notes in this gist](https://gist.github.com/romgrk/28a6106c56c05f5f932d5d53dbcecfef)). That's of course a general solution to selector-based reactivity's issues, but yours could provide more immediate gains. I'm wary though of adding quick workarounds, the codebase already has a fair amount of tech debt to resolve.

lauri865 commented:
> I've also observed that issue, and I've mentionned that we could get rid of the perf cost of running all the selectors during scrolling by switching from selector-based reactivity to event-based reactivity ([conceptual notes in this gist](https://gist.github.com/romgrk/28a6106c56c05f5f932d5d53dbcecfef)). That's of course a general solution to selector-based reactivity's issues, but yours could provide more immediate gains. I'm wary though of adding quick workarounds, the codebase already has a fair amount of tech debt to resolve.
> 
> That's also why `getCellParams` was a single bag-of-data selector previously, it allowed to keep the number of selector-listeners as low as possible.

Thanks for the additional context, sounds like an interesting approach. I thought about alternative approaches as well, but after some quick measurements, I was more leaning towards the number of selectors not being an issue in and of itself – they're pretty performant. It's only an issue when the state updates a) with high frequency and b) we're rendering based on each state change / where every last ms counts. The only two areas I can think of are `renderContext` and potentially `dimensions` (if not throttled, but even that shouldn't be much of an issue anymore).

So, that leaves only `renderContext` (at least to my current understanding, let me know if you can think of anything else). Unless we can measurably show an issue, then maybe it's not worth refactoring too much of the code base, only to measurably gain in one pretty contained use case.

cherniavskii commented on packages/x-data-grid-pro/src/hooks/features/treeData/gridTreeDataUtils.ts:
```diff
@@ -103,7 +102,9 @@ export const filterRowTreeFromTreeData = (
 
-    filteredRowsLookup[node.id] = shouldPassFilters;
+    if (!shouldPassFilters) {
+      filteredRowsLookup[node.id] = false;
```
I spotted 2 places like this where we were still assigning `true` values, so I narrowed the type in [`3449782` (#15648)](https://github.com/mui/mui-x/pull/15648/commits/34497825839833e0ce866797bc8c1b93686197ae)

lauri865 replied:
Good catch, thanks. Fixed a few checks as well on that front, and tests are passing again for all cases.

lauri865 commented:
Since streamlining state updates made `detailPanel` tests fail, I was forced to fix some measurement propagation issues on that front as well. Now there's no layout shifts when opening up detail panels anymore, and further reduced state updates:

Before:

https://github.com/user-attachments/assets/9e6585c6-defa-41e0-8606-de4e2e0e6ede

After:

https://github.com/user-attachments/assets/42f70ccf-b04c-45b1-948c-c0f804698e77

cherniavskii commented on packages/x-data-grid/src/hooks/features/filter/gridFilterState.ts:
```diff
@@ -67,4 +73,4 @@ export type GridFilteringMethodValue = Omit<GridFilterState, 'filterModel'>;
  * If a row is not registered in this lookup, it is visible.
  */
-export type GridVisibleRowsLookupState = Record<GridRowId, boolean>;
+export type GridVisibleRowsLookupState = Record<GridRowId, false>;
```
Same change for `visibleRowsLookup` as previously with `filteredRowsLookup`.

cherniavskii commented on packages/x-data-grid/src/hooks/utils/useGridSelector.ts:
```diff
@@ -102,20 +105,38 @@ export const useGridSelector = <Api extends GridApiCommon, Args, T>(
 
-  useOnMount(() => {
-    return apiRef.current.store.subscribe(() => {
+  const subscribe = useEventCallback(() => {
```
`useEventCallback` seemed cleaner to me

lauri865 replied:
Adds an unnecessary useLayouEffect though that runs a function on every render. @romgrk suggested the previous approach with emtpy array and muting eslint to keep this code path maximally optimised.

MBilalShafi commented on packages/x-data-grid/src/hooks/features/filter/gridFilterState.ts:
```diff
@@ -20,7 +26,7 @@ export interface GridFilterState {
    * This is the equivalent of the `visibleRowsLookup` if all the groups were expanded.
    */
-  filteredRowsLookup: Record<GridRowId, boolean>;
+  filteredRowsLookup: Record<GridRowId, false>;
```
Since we don't store `true` values now, does it make sense to simplify the type further? It could be slightly better performance wise. (same for the visible rows, column visibility model, etc)
```suggestion
  unfilteredRows: Set<GridRowId>;
```

cherniavskii replied:
Changing `Record` to `Set`/`Map` won't necessarily be faster – see https://github.com/mui/mui-x/pull/9120#discussion_r1215040795

romgrk replied:
See https://github.com/mui/mui-x/pull/10068#issuecomment-1743654847 and https://github.com/mui/mui-x/pull/10068#discussion_r1355431915. I think the change is good and has potential for improvement, but I didn't have the bandwidth to investigate it more.

romgrk commented on packages/x-data-grid-pro/src/components/GridPinnedRows.tsx:
```diff
@@ -20,18 +19,17 @@ export function GridPinnedRows({ position, virtualScroller }: GridPinnedRowsProp
       lastRowIndex: rows.length,
-      firstColumnIndex: renderContext.firstColumnIndex,
-      lastColumnIndex: renderContext.lastColumnIndex,
+      firstColumnIndex: undefined,
+      lastColumnIndex: undefined,
```
If this value isn't used, can we use `-1` as an empty value marker instead? It will [avoid changing the shape](https://romgrk.com/posts/optimizing-javascript#2-avoid-different-shapes) of those objects.

romgrk commented on packages/x-data-grid/src/components/GridScrollArea.tsx:
```diff
@@ -62,21 +65,52 @@ const GridScrollAreaRawRoot = styled('div', {
+  const apiRef = useGridApiContext();
+  const [dragging, setDragging] = React.useState<boolean>(false);
+
+  const handleColumnHeaderDragStart = useEventCallback(() => {
+    setDragging(true);
+  });
+
+  const handleColumnHeaderDragEnd = useEventCallback(() => {
+    setDragging(false);
+  });
+
+  useGridApiEventHandler(apiRef, 'columnHeaderDragStart', handleColumnHeaderDragStart);
+  useGridApiEventHandler(apiRef, 'columnHeaderDragEnd', handleColumnHeaderDragEnd);
```
`useGridApiEventHandler` doesn't need memoized functions (with `useEventCallback`), it already replicates the same logic that `useEventCallback` implements.

romgrk commented on packages/x-data-grid/src/internals/selectors/dimensionSelectors.ts:
This file should probably be under the `features/dimensions` folder, to keep with existing conventions. You can add an `export { ... } from '.../features/dimensions/dimensionSelectors'` where appropriate to expose these from `@mui/x-data-grid/internals`.

romgrk commented on packages/x-data-grid/src/internals/selectors/dimensionSelectors.ts:
```diff
@@ -0,0 +1,43 @@
+
+export const gridRowHeightSelector = (state: GridStateCommunity) => state.dimensions.rowHeight;
+
+export const gridDimensionsColumnsTotalWidthSelector = (state: GridStateCommunity) =>
+  state.dimensions.columnsTotalWidth;
```
You said there is a name conflict with another selector called `gridColumnsTotalWidthSelector`. Are the two selectors returning the same value, calculated differently? If so, would it be appropriate to remove one of them?

cherniavskii replied:
> Are the two selectors returning the same value, calculated differently? If so, would it be appropriate to remove one of them?

No, the original `gridColumnsTotalWidthSelector` calculates the value, and that value change triggers the dimensions state update and gets stored in that state.
AFAIU @lauri865 used `gridDimensionsColumnsTotalWidthSelector` instead because it guarantees that the dimensions state is up to date.

I've moved the original selector to `useGridDimensions`, it's only used internally now.
I've also renamed `gridDimensionsColumnsTotalWidthSelector` to `gridColumnsTotalWidthSelector` and used `createSelector` to keep the signature of the original selector.

@lauri865 Does this make sense to you?

lauri865 replied:
Exactly. I believe your change makes sense, as nothing else probably depends on the calculated size other than `useGridDimensions`, it makes sense to localise it.

romgrk commented on packages/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx:
```diff
@@ -378,14 +398,25 @@ export const useGridVirtualScroller = () => {
       rows?: GridRowEntry[];
       position?: GridPinnedRowsPosition;
-      renderContext?: GridRenderContext;
+      renderContext?:
+        | GridRenderContext
+        | (Pick<GridRenderContext, 'firstRowIndex' | 'lastRowIndex'> & {
+            firstColumnIndex: undefined;
+            lastColumnIndex: undefined;
+          });
```
We could also use `-1` for the type.

romgrk reviewed (approved):
LGTM beyond the small comments above.

arminmeh commented on packages/x-data-grid/src/hooks/features/dimensions/gridDimensionsSelectors.ts:
```diff
@@ -1,3 +1,52 @@
+  (dimensions) => dimensions.columnsTotalWidth,
+);
+
+export const gridRowHeightSelector = (state: GridStateCommunity) => state.dimensions.rowHeight;
```
@lauri865 
any particular reason why all these selectors are not using `gridDimensionsSelector` as an input?

lauri865 replied:
What would it be solving to wrap them in another function call? Just went with my gut feel tbh - simple selectors, simpler to write plainly. Localized to a single file, enforced by typescript.

romgrk replied:
They should probably be transformed to something like `const xxxSelector = createRootSelector(state => state.dimensions.xxx)` to keep them readable.
